pageserver/src/tenant/layer_map.rs

```diff
@@ -275,10 +275,9 @@ where
     /// Helper function for BatchedUpdates::insert_historic
     ///
     pub(self) fn insert_historic_noflush(&mut self, layer: Arc<L>) {
-        self.historic.insert(
-            historic_layer_coverage::LayerKey::from(&*layer),
-            Arc::clone(&layer),
-        );
+        let key = historic_layer_coverage::LayerKey::from(&*layer);
+        assert!(!self.historic.contains(&key));
```

Comment:
Here  your and @koivunej  opinions are opposite. 
There was original `error!` then I replace it with `assert!`
If there is no any valid reasons for storing duplicates layer (and at this moment I do not know one),
then IMHO assert is more preferable.

Reply:
`error!` would just print a message to the log, I agree that's not good either. We don't want to ever have duplicates in the layer map. I meant this should return an error, i.e. `bail!(...)` or similar.

Reply:
But what is the difference between assert! and bail! ?
Both cause panic which been unhandled should terminate current task.

Reply:
> But what is the difference between assert! and bail! ?
> Both cause panic which been unhandled should terminate current task.

No, they are very different. `assert!(condition)` is equivalent to:
```
if !condition {
   panic!();
}
```
while `bail!(message)` is equivalent to:
```
return Err(anyhow:Error(message))
```
`bail!` doesn't panic.

Reply:
Also, if I will use bail!, then I will have to propagate this error upward - change profile of several existed insert* functions.
I wonder why do we need it if we are not going to handle this error in any case?

Reply:
> I wonder why do we need it if we are not going to handle this error in any case?

assert or `Err`[or] it will be difficult on the current callsites. Note: I think my thread was on the previous PR and this PR as well: https://github.com/neondatabase/neon/pull/3869#discussion_r1146108609

Perhaps even unwrapping on the callsite rather than to undo everything changed in the batch for example (there's no undo there currently it will apply changes via drop which will happen on panic or `?`)... This is a tricky problem. I doubt there are any single-line solutions.

We could make `BatchedUpdates` be only apply/flush on-demand (instead on drop), then move all undoable code to happen after this on-demand apply/flush. Then we'd have nothing to undo, and we could use the `?`.

Reply:
Hmmm...
It seems to me that assumption is that presence of duplicate layers is a bug.
It is not some normal situation which we should handle.
So I do not think why do we need some more sophisticated logic here, handle batch updates,...
My option is  handling of this error should satisfy three requirements:
1. It can be easily detected (no need to inspect all logs for presence of this error message)
2. Log message should contain enough information to understand the reason of the problem (stack trace). panic! is so more preferable, s it is dumps stack by default, while unhandled error requires RUST_BACKTRACE=1
3. It should not affect other clients (so should not cause crash of the whole pageserver is such bugs happens).

The simplest approach with assert satisfies all this requirements.
I do not understand why do we need something else.

Reply:
re: my earlier comment, I had understood this was a discussion about the callsites of `insert_historic`; I think there used to be one assert earlier, and I suggested this `HashSet::insert(&mut self, val: T) -> bool` alike api, then a fix for the original impl and now it seems the `assert!` is back inside the `insert_historic` and it's now gone from the callsites. Hiding my earlier response. The link should be to the other PR.

Reply:
I am sorry for confusion.
So the primary intention was to add check for duplicated layers.
Originally I tried to do it at as low level as possible - in LayerCoverage.
But then I found out that it cause failure of some tests. This is why I decided to do in `create_image_layers` when image layers are generate. But then I review this decision and investigate more this test failure which actually indicates real errors.
So now this PR includes not only test for duplicates but also fixes of this errors.